### PR TITLE
Appropriate cluster sizing

### PR DESCRIPTION
**Pull request by dylanbannon**
Users now have the ability to specify a lot of optoins during configuration, including GPU pool sizes and GPU types. Default values for these options are, more or less, sane.

This addresses issues #55 and #57.

Oh, sweet. It also addresses issue #30.

### CODE REVIEW

**Comment by willgraf** [comment]
Looks good.  Is there a way to give the user options for the GPU node values?

If they are just entering a text field, anyone trying to change the value would likely get it wrong.  I'm a bit more knowledgeable than the target users, but I had no idea what other value to put in the text box.

File: conf/tasks/Makefile.kops [hunk 1]
```diff
@@ -1,6 +1,13 @@
 SSH_KEY ?= $(CACHE_PATH)/id_rsa
 SSH_PUBLIC_KEY ?= $(SSH_KEY).pub
 
+## Prep the kiosk for cluster creation
+kops/prep:
+	@echo "Prepping the kiosk for cluster creation..."
+	apk add gettext
```
Comment (osterman) [inline conf/tasks/Makefile.kops]: `gettext` should already be in `geodesic`, but either way, move this to the `Dockerfile`

File: conf/tasks/Makefile.kops [hunk 2]
```diff
@@ -43,7 +50,8 @@ kops/create/aws:
 	@echo "Briefly restarting the AWS cluster in order to add some necessary permissions to it..."
 	@kops get cluster -o yaml > /tmp/cluster-template.yaml
 	@yq merge -x /tmp/cluster-template.yaml patches/aws.yaml > /tmp/cluster.yaml
-	@kops replace -f /tmp/cluster.yaml
+	@cat /tmp/cluster.yaml | envsubst > /tmp/cluster-envsubst.yaml
```
Comment (osterman) [inline conf/tasks/Makefile.kops]:
```suggestion
	@envsubst <  /tmp/cluster.yaml > /tmp/cluster-envsubst.yaml
```

File: conf/tasks/Makefile.kops [hunk 3]
```diff
@@ -43,7 +50,8 @@ kops/create/aws:
...
+	@kops replace -f /tmp/cluster-envsubst.yaml
 	@kops update cluster --yes || echo "Restart and update didn't happen."
 	@echo "Permission addition finished."
 	@echo " "
```
Comment (osterman) [inline conf/tasks/Makefile.kops]:
```suggestion
	@echo
```

File: conf/tasks/Makefile.kops [hunk 4]
```diff
@@ -57,7 +65,8 @@ kops/create/gpu-nodes:
 	@if [[ -z "${GPU_NODES_EXIST}" ]]; then \
 		kops create ig gpu-nodes -o yaml --dry-run > /tmp/gpu-nodes-template.yaml; \
 		yq merge -x /tmp/gpu-nodes-template.yaml patches/gpu-nodes.yaml > /tmp/gpu-nodes.yaml; \
-		kops create -f /tmp/gpu-nodes.yaml; \
+		cat /tmp/gpu-nodes.yaml | envsubst > /tmp/gpu-nodes-envsubst.yaml; \
```
Comment (osterman) [inline conf/tasks/Makefile.kops]:
```suggestion
	envsubst <  /tmp/gpu-nodes.yaml  > /tmp/gpu-nodes-envsubst.yaml; \
```

File: scripts/menu.sh [hunk 5]
```diff
@@ -123,6 +127,14 @@ function configure_gke() {
...
   export CLUSTER_NAME=$(inputbox "Deepcell" "Cluster Name" "${CLUSTER_NAME:-deepcell}")
   export GKE_BUCKET=$(inputbox "Deepcell" "Bucket Name" "${GKE_BUCKET}")
+  export GKE_COMPUTE_REGION=$(inputbox "Google Cloud" "Compute Region" "us-west1")
+  export GKE_COMPUTE_ZONE=$(inputbox "Google Cloud" "Compute Zone" "us-west1-b")
+  export GKE_MACHINE_TYPE=$(inputbox "Google Cloud" "Node (non-GPU) Type" "n1-standard-2")
+  export GPU_TYPE=$(inputbox "Google Cloud" "GPU type" "nvidia-tesla-p100")
+  export GPU_PER_NODE=$(inputbox "Google Cloud" "GPUs per GPU Node" "1")
+  export GPU_MACHINE_TYPE=$(inputbox "Google Cloud" "GPU Node Type" "n1-standard-4")
+  export GPU_NODE_MIN_SIZE=$(inputbox "Google Cloud" "Minimum Number of GPU Nodes" "0")
+  export GPU_NODE_MAX_SIZE=$(inputbox "Google Cloud" "Maximum Number of GPU Nodes" "4")
```
Comment (osterman) [inline scripts/menu.sh]: if you pass `4` explicitly, then if the user goes back to re-edit their cluster, it won't use their last known value. 

Suggest instead writing it `"${GPU_NODE_MAX_SIZE:-4}"`

Comment (osterman) [inline scripts/menu.sh]:
```suggestion
  export GPU_NODE_MAX_SIZE=$(inputbox "Google Cloud" "Maximum Number of GPU Nodes" "${GPU_NODE_MAX_SIZE:-4}")
```

Comment (osterman) [reply scripts/menu.sh]: ...then repeat for the others

File: conf/tasks/Makefile.kops (hunk 1 above)
Comment (osterman) [reply conf/tasks/Makefile.kops]: `read` function is actually a bash internal.  `gettext` is for `envsubst`

Comment (osterman) [reply conf/tasks/Makefile.kops]: hrm... though there is a `/usr/bin/read` too... 😕

**Comment by dylanbannon** [comment]
Addressing, @willgraf's comment, I've pushed some commits that turn GPU selection into a pre-populated radio button selection screen.

File: scripts/menu.sh (hunk 5 above)
Comment (dylanbannon) [reply scripts/menu.sh]: I didn't even think of this. Done.

Comment (dylanbannon) [reply scripts/menu.sh]: Also done.

**Comment by dylanbannon** [comment]
@osterman I've addressed all your comments but I did some of them via commits, rather than Github.

File: scripts/menu.sh [hunk 6]
```diff
@@ -107,6 +125,26 @@ function configure_aws() {
   export AWS_SECRET_ACCESS_KEY=$(inputbox "Amazon Web Services" "AWS Secret Key" "${AWS_SECRET_ACCESS_KEY}")
   export AWS_S3_BUCKET=$(inputbox "Amazon Web Services" "AWS S3 Bucket Name" "${AWS_S3_BUCKET}")
   export NAMESPACE=$(inputbox "Deepcell" "Cluster Name" "${NAMESPACE}")
+  export MASTER_MACHINE_TYPE=$(inputbox "Amazon Web Services" "Master Node Machine Type" "${MASTER_MACHINE_TYPE:-m4.large}")
+  export NODE_MACHINE_TYPE=$(inputbox "Amazon Web Services" "Worker Nodes Machine Type" "${NODE_MACHINE_TYPE:-m4.large}")
+
+  # let's just hardcode the menu, since all instance types are apparently available in all regions
+  # and there's no built-in way to list type in the aws-cli
+  local gpu_types="g3s.xlarge 1GPU OFF
```
Comment (willgraf) [inline scripts/menu.sh]: screen shot 2018-11-15 at 3 28 46 pm

Comment (dylanbannon) [reply scripts/menu.sh]: Looks like I should have done more testing...

Comment (willgraf) [reply scripts/menu.sh]: Looks like the typo was indeed on the AWS website.  Changing `g3s.xlarge` to `g3.xlarge` allowed the cluster to be created.

**Comment by willgraf** [comment]
Encountered GPU node scaling issue on both AWS and GKE with this PR.

Both deployments had a `tf-serving-deployment` stuck in `Pending` for over 30 minutes (when the `check_tf_liveness` times out). There was no evidence that a GPU node was being brought up for the cluster.

**Comment by dylanbannon** [comment]
Following up on @willgraf, I saw on GKE that GPU scaling was working for training jobs, but not for prediction jobs. I did not test anything on AWS.

Is GPU scaling working for prediction jobs on the `master` branch? If so, we should merge master into this branch and figure out what the issue is during the merge.

**Comment by willgraf** [comment]
> Following up on @willgraf, I saw on GKE that GPU scaling was working for training jobs, but not for prediction jobs. I did not test anything on AWS.
> 
> Is GPU scaling working for prediction jobs on the `master` branch? If so, we should merge master into this branch and figure out what the issue is during the merge.

Yes, GPU scaling is working on the master branch (at least for GKE):
screen shot 2018-11-16 at 8 44 05 am

**Comment by dylanbannon** [comment]
Yeah, @willgraf, I'm seeing prediction-related GPU scaling for AWS on the `master` branch. I'm going to merge `master` into `appropriate-cluster-sizing` in the hopes of fixing this.

**Comment by dylanbannon** [comment]
I played around with this and found GKE had no problems granting GPUs when we chose to use K80 GPUs. When we choose P100 GPUs, we get GPU node pool scale up for training jobs, but not for prediction.

To address this, I'm going to set the default GPU type to be K80 on both GKE and AWS (specifically, "p2.xlarge" on AWS) and open an issue surrounding this GPU-type-specific failure.

**Comment by dylanbannon** [comment]
Ok, last comment and then we're merging.

@willgraf was concerned a few comments ago that GPU scaling wasn't working correctly on this branch. He's correct about that, but
1) this branch doesn't break any GPU scaling functionality that previously existed (or currently exists) in `master`
and
2) this branch does accomplish its originally-stated goal of allowing users to specify a bunch of options and implementing sane defaults for those options.

On the basis of that, I'm going to advocate for a merging of this branch into `master`.


Now, as for the GPU scaling...
I tested this a lot and here's what I found:
On `master` (note that both clusters are hardcoded to use K80 GPUs):
GKE-predict-> works
AWS-predict -> works
On this branch:
GKE-K80-predict -> works
AWS-K80-predict -> works
GKE-K80-train -> works
AWS-K80-train -> doesn't work
GKE-P100-predict -> doesn't work
GKE-P100-train -> works

This shows that everything that works in master works in this branch, and that there are two different cases (of those tested here) that fail. I'm going to open one issue about those cases, but I suspect they have different causes, since they involve different clouds, different GPUs, and different cluster functionality.

**Pull request merged**